add docs about the default storage class

ggilmore opened the pull request:
See https://github.com/sourcegraph/deploy-sourcegraph/issues/84

ggilmore commented:
I added a stub storage class (we can't specify a general one - the fields are hosting provider specific). 

`./kubectl-apply-all` will say that the storage class is invalid if you don't fill out the stub before applying the configuration (but the rest of the deployment will still succeed). Maybe we can create a storage class with a different name so that `kubectl` will really road-block you if you didn't specify the field?

nicksnyder commented on base/default.StorageClass.yaml:
```diff
@@ -0,0 +1,14 @@
+kind: StorageClass
```
I don't think this should be included here because it is invalid. I think it would be sufficient to just inline this fragment into the README itself.

ggilmore commented:
@nicksnyder PTAL. I still need to verify that if you batch `kubectl apply` files, that it's guaranteed that the storage class is  config applied _before_ the PVCs are created.

nicksnyder commented on docs/configure.md:
````diff
@@ -377,7 +377,34 @@ See [the official documentation](https://kubernetes.io/docs/concepts/configurati
+
+1. Create a stub base/default.StorageClass.yaml.
+
+   ```yaml
````
can we provide a concrete example for GCP at least, ideally AWS too?

ggilmore replied:
Actually, this might not be strictly true. According to https://kubernetes.io/docs/concepts/storage/storage-classes/#gce, not specifying a `zone` will cause requests to round-robin between all the different zones that your nodes reside in (which might not be desirable).

nicksnyder replied:
I still think an example helps.

For example you could have a commented out section that shows what a real GCP example would look like.

chrismwendt replied:
The Azure storage class must be `managed-premium` https://github.com/sourcegraph/sourcegraph/issues/10795

nicksnyder commented on docs/install.md:
````diff
@@ -30,6 +30,8 @@
    git clone https://github.com/sourcegraph/deploy-sourcegraph && cd deploy-sourcegraph && git checkout ${VERSION}
    ```
 
+1. Configure the cluster's default storage class by reading through "Configure a storage class" in [docs/configure.md](../configure.md#configure-a-storage-class).
````
make sure this gets applied first

ggilmore commented:
@nicksnyder PTAL

nicksnyder commented on docs/configure.md:
```diff
@@ -377,7 +377,52 @@ See [the official documentation](https://kubernetes.io/docs/concepts/configurati
 ## Configure a storage class
 
-Sourcegraph relies on the default storage class of your cluster. If your cluster does not have a default storage class or if you wish to use a different storage class for Sourcegraph, then you need to update all PersistentVolumeClaims with the name of the desired storage class.
+**Note:** The storage class configuration steps should be done _before_ running `./kubectl-apply-all.sh` for the first time.
```
probably worth explaining what happens if you don't do this

ggilmore commented:
@nicksnyder PTAL

ggilmore commented:
@nicksnyder PTAL, there are more instructions that are needed so that the previously existing SC's are marked as non-default

ggilmore commented on docs/configure.md:
```diff
@@ -377,7 +377,70 @@ See [the official documentation](https://kubernetes.io/docs/concepts/configurati
+   #!/bin/bash
+
+   # Mark the existing storage classes in the cluster as non-default
+   kubectl get sc -o json | jq --raw-output ".items | map(select(.metadata.name)) | .[] | \"kubectl patch sc -p '{\\\"metadata\\\":{\\\"annotations\\\":{\\\"storageclass.kubernetes.io/is-default-class\\\":\\\"false\\\"}}}' \\(.metadata.name)\"" | bash
```
Actually, this won't work if you have a storage class with the `"storageclass.beta.kubernetes.io/is-default-class` annotation. I'm trying to figure out a jq script that can handle both cases.

nicksnyder commented:
If we are lifting the SSD requirement into the main install instructions, then maybe we should not mess with default storage class entirely.

We can pre-emptively use a custom storage class name in the base yaml (e.g. sourcegraph-ssd) which will require the creation of a non-default storage class with that name.

No only does this seem simpler, it more forcefully nudges customers to use ssds.

beyang commented on docs/install.md:
````diff
@@ -30,6 +30,8 @@
    git clone https://github.com/sourcegraph/deploy-sourcegraph && cd deploy-sourcegraph && git checkout ${VERSION}
    ```
 
+1. Configure the `sourcegraph` storage class for the cluster by reading through "Configure a storage class" in [docs/configure.md](../configure.md#configure-a-storage-class).
````
`["Configure a storage class"](../configure.md#configure-a-storage-class).`

The link text should correspond to what's linked to.

ggilmore commented:
@nicksnyder @beyang PTAL, I added migration instructions concerning the storage classes and added snippets for the different hosting providers for using SSDs (I'm not sure if the value are correct for AWS and Azure - this is only from looking at the documentation). 

@beyang - @nicksnyder and I talked about this, but configuring SSDs this way introduces a breaking change since the storage class name is newly specified.
